storage/innobase/buf/buf0buf.cc

```diff
@@ -1394,17 +1394,14 @@ inline bool buf_pool_t::withdraw_blocks()
 
 		/* reserve free_list length */
 		if (UT_LIST_GET_LEN(withdraw) < withdraw_target) {
-			buf_flush_LRU(
-				std::max<ulint>(withdraw_target
-						- UT_LIST_GET_LEN(withdraw),
-						srv_LRU_scan_depth),
-				true);
-			mysql_mutex_unlock(&buf_pool.mutex);
-			buf_dblwr.flush_buffered_writes();
-			mysql_mutex_lock(&buf_pool.flush_list_mutex);
-			buf_flush_wait_LRU_batch_end();
-			mysql_mutex_unlock(&buf_pool.flush_list_mutex);
-			mysql_mutex_lock(&buf_pool.mutex);
+			try_LRU_scan = false;
+			mysql_mutex_lock(&flush_list_mutex);
+			mysql_mutex_unlock(&mutex);
```

Comment:
For the stack traces of hang that I posted to [MDEV-33613](https://jira.mariadb.org/browse/MDEV-33613), for the commit 1e0e0e27b862fc0907f9c9599e467edfc4c9b3c4, we can see a hang exactly here, in the `my_cond_wait()`:
```c++
		const bool wake = buf_pool.need_LRU_eviction();
		mysql_mutex_unlock(&buf_pool.mutex);
		mysql_mutex_lock(&buf_pool.flush_list_mutex);
		const auto n_flush = buf_pool.n_flush();
		if (wake && !buf_pool.page_cleaner_active()) {
			buf_pool.page_cleaner_wakeup(true);
		}
		mysql_mutex_unlock(&buf_pool.flush_list_mutex);
		mysql_mutex_lock(&buf_pool.mutex);
		if (!n_flush) {
			goto not_found;
		}
		if (!buf_pool.try_LRU_scan) {
			my_cond_wait(&buf_pool.done_free,
				     &buf_pool.mutex.m_mutex);
		}
```
By contiguously holding the `buf_pool.mutex` while waiting for `buf_pool.flush_list_mutex` we prevent the scenario. I reliably reproduced this on my system for e8b28ea3ac772cbc96c74d002c3d0e305a79d2f4 and not after the merge of this fix (30065570d024ac4ffb65b7541b102233aade9180).

This reminds me of [MDEV-30180](https://jira.mariadb.org/browse/MDEV-30180). There, a fix was similar: acquire and hold `buf_pool.mutex` before waiting for `buf_pool.flush_list_mutex`. The scenario was different from this one: We wanted to ensure that a pending `buf_page_write_complete()` will complete (it could be holding `buf_pool.mutex`), which in turn would allow `buf_flush_page_cleaner()` to make some progress. I mention this because I remembered those earlier ‘unfair scheduling’ type hangs and was worrying if this change could be causing more of that. Ultimately, those types of hangs were fixed by 3613fb2aa85abd67313bd974253755358f430f3c and f8c88d905b44bffe161158309e9acc25ad3691aa.

Reply:
The condition variable `buf_pool.done_free` is associated with `buf_pool.mutex`. If we release `buf_pool.mutex` while waiting for `buf_pool.flush_list_mutex`, we open up opportunities for more interleaved executions that are hard (at least for me) to reason about.

I will check which other threads could be holding both mutexes at the same time. It should be rare.

Reply:
Perhaps you are confusing with other flow ? In current case it is associated with flush_list_mutex.m_mutex for done_flush_list. Please see the next lines.

my_cond_wait(&done_flush_list, &flush_list_mutex.m_mutex);

Reply:
If you are thinking about other independent execution paths dealing with buf_pool.done_free I don't see how that would matter. Note that we are releasing the buffer pool mutex immediately afterwards and finally would be releasing the flush list mutex while waiting.

Reply:
Oh, sorry, your comment was about buffer pool resizing, while I thought that it was about `buf_LRU_get_free_block()`. Sorry. I see no problem to release `buf_pool.mutex` earlier there. An untimed wait for `buf_pool.done_flush_list` under `buf_pool.flush_list_mutex` should work there. Should it hang, that is something that would be addressed in #3107 ([MDEV-29445](https://jira.mariadb.org/browse/MDEV-29445)) at some later point of time.

Reply:
Addressed in d157a3c0c6557b8db8ce2ff6764e2fad0cf21821.